libgadget/run.c

```diff
@@ -386,14 +387,21 @@ run(int RestartSnapNum)
         const int NeutrinoTracer =  All.HybridNeutrinosOn && (atime <= All.HybridNuPartTime);
         const double rho0 = All.CP.Omega0 * 3 * All.CP.Hubble * All.CP.Hubble / (8 * M_PI * All.CP.GravInternal);
 
-        if(All.TreeGravOn) {
+        /* Build the force tree for tree gravity.*/
+        ForceTree Tree = {0};
+        int anygravactive = MPIU_Any(Act.NumActiveGravity, MPI_COMM_WORLD);
+
+        if(is_PM || (All.TreeGravOn && anygravactive))
+            force_tree_rebuild(&Tree, ddecomp, HybridNuGrav, !pairwisestep && All.TreeGravOn, All.OutputDir);
+
+        if(All.TreeGravOn && (is_PM || anygravactive)) {
```

Comment:
This looks puzzling -- how is this condition related to the one in line 394? Could you add a comment on when the first is True but the second is False? And looks like we will crash if the first is False but the second is True?

Reply:
Yes, this is indeed confusing. The first is true and the second false when tree gravity is off. Then we still need a tree to do the PM. The first being false but the second being true should be impossible (I think): crashing is a legit thing to do there. I will add a comment.

Reply:
"still need a tree to do the PM": it sounds like even after the GasTree split, the remaining gravTree can still be refactored into two -- the spatial indexing for PM, and the "moments" for short-range grav; in the future.

Reply:
Yes such a split would be a good idea. We could save some memory for the gasTree, and reduce peak memory usage during density(). Just needs some messing about because C does not do generics :) It is not worth optimising the case where gravTree is disabled and density is on to avoid an extra rebuild because we almost never run in that mode.